go/teams/treeloader.go

```diff
@@ -10,204 +10,248 @@ import (
-		mctx.Debug("loadTeamTreeMembershipsRecursive: short-circuiting load due to failure")
-		return nTeamsLoaded
+		mctx.Debug("Treeloader.loadRecursive: short-circuiting load due to failure: %+v", result)
+		return expectedCount
 	}
 
 	eg, ctx := errgroup.WithContext(mctx.Ctx())
 	mctx = mctx.WithContext(ctx)
 	// Load ancestors
 	if np == nodePositionTarget && !teamName.IsRootTeam() {
 		eg.Go(func() error {
-			incr := loadTeamAncestorsMemberships(mctx, teamID, teamName, uv, converter)
-			mctx.Debug("loadTeamTreeMembershipsRecursive: loaded %d teams from ancestors", incr)
-			atomic.AddInt32(&nTeamsLoaded, incr)
+			incr := l.loadAncestors(mctx, teamID, teamName)
+			mctx.Debug("Treeloader.loadRecursive: loaded %d teams from ancestors", incr)
+			atomic.AddInt32(&expectedCount, incr)
 			return nil
 		})
 	}
 	// Load subtree
 	// Because we load parents before children, the child's load can use the cached parent's team
 	// so we only make one team/get per team.
-	for _, idAndName := range node.chain().ListSubteams() {
+	for _, idAndName := range team.chain().ListSubteams() {
 		idAndName := idAndName
 		// This is unbounded but assuming subteam spread isn't too high, should be ok.
 		eg.Go(func() error {
-			incr := loadTeamTreeMembershipsRecursive(
-				mctx, idAndName.Id, idAndName.Name, uv, nodePositionChild, converter,
-			)
-			mctx.Debug("loadTeamTreeMembershipsRecursive: loaded %d teams from subtree", incr)
-			atomic.AddInt32(&nTeamsLoaded, incr)
+			incr := l.loadRecursive(mctx, idAndName.Id, idAndName.Name)
+			mctx.Debug("Treeloader.loadRecursive: loaded %d teams from subtree", incr)
+			atomic.AddInt32(&expectedCount, incr)
 			return nil
 		})
 	}
 	// Should not return any errors since we did error handling ourselves
 	_ = eg.Wait()
 
-	return nTeamsLoaded
+	return expectedCount
 }
 
-func loadTeamAncestorsMemberships(mctx libkb.MetaContext, teamID keybase1.TeamID,
-	targetTeamName keybase1.TeamName, uv keybase1.UserVersion,
-	converter TreeloaderStateConverter) (nTeamsLoaded int32) {
-	defer mctx.TraceTimed(fmt.Sprintf("loadTeamAncestorsMemberships(%s, %s)", targetTeamName, uv),
-		func() error { return nil })()
+func (l *Treeloader) loadAncestors(mctx libkb.MetaContext, teamID keybase1.TeamID,
+	teamName keybase1.TeamName) (expectedCount int32) {
+	defer mctx.TraceTimed(fmt.Sprintf("Treeloader.loadAncestors"), func() error { return nil })()
 
-	handleAncestor := func(t keybase1.TeamSigChainState, teamName keybase1.TeamName) error {
-		res := converter.SigchainStateToTeamTreeMembership(mctx, &t, uv)
-		s, err := res.S()
-		if err != nil {
-			return err
-		}
-		// Short-circuit ancestor load if this resulted in an error (though it shouldn't except for
-		// in testing), to keep symmetry with behavior if the ancestor team load failed.
+	handleAncestor := func(t keybase1.TeamSigChainState, ancestorTeamName keybase1.TeamName) error {
+		result := l.Converter.ProcessSigchainState(mctx, ancestorTeamName, &t)
+		s, _ := result.S()
```

Comment:
why ignore the error here? I agree it should never happen, but still....

Reply:
I unignored the error here, but I do this in another place in this file where it's more annoying to handle the errorpath. I could add a panic maybe? (ideally it would be guaranteed by the protocol that this data isn't inconsistent...)

Reply:
My instinct would be to go for a panic, so that if an error happens we know what is going on immediately (instead of having other weird errors down the line. However this is risky in case it does fail, so I would ask others how they deal with this. Do not feel strongly either way.